File: app/src/main/java/dev/ragnarok/fenrir/db/impl/DatabaseStorage.java
package dev.ragnarok.fenrir.db.impl;

import static dev.ragnarok.fenrir.util.Utils.safeCountOf;

import android.content.ContentProviderOperation;
import android.content.ContentValues;
import android.database.Cursor;
import android.net.Uri;
import android.provider.BaseColumns;

import androidx.annotation.NonNull;

import java.util.ArrayList;
import java.util.List;

import dev.ragnarok.fenrir.db.MessengerContentProvider;
import dev.ragnarok.fenrir.db.column.CountriesColumns;
import dev.ragnarok.fenrir.db.interfaces.IDatabaseStore;
import dev.ragnarok.fenrir.db.model.entity.CountryEntity;
import dev.ragnarok.fenrir.util.Objects;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Single;


public class DatabaseStorage extends AbsStorage implements IDatabaseStore {

    DatabaseStorage(@NonNull AppStorages base) {
        super(base);
    }

    @Override
    public Completable storeCountries(int accountId, List<CountryEntity> dbos) {
        return Completable.create(emitter -> {
            Uri uri = MessengerContentProvider.getCountriesContentUriFor(accountId);

            ArrayList<ContentProviderOperation> operations = new ArrayList<>(dbos.size() + 1);
            operations.add(ContentProviderOperation.newDelete(uri).build());

            for (CountryEntity dbo : dbos) {
                ContentValues cv = new ContentValues();
                cv.put(BaseColumns._ID, dbo.getId());
                cv.put(CountriesColumns.NAME, dbo.getTitle());

                operations.add(ContentProviderOperation.newInsert(uri)
                        .withValues(cv)
                        .build());
            }

            getContentResolver().applyBatch(MessengerContentProvider.AUTHORITY, operations);
            emitter.onComplete();
        });
    }

    @Override
    public Single<List<CountryEntity>> getCountries(int accountId) {
        return Single.create(emitter -> {
            Uri uri = MessengerContentProvider.getCountriesContentUriFor(accountId);

            Cursor cursor = getContentResolver().query(uri, null, null, null, null);
            List<CountryEntity> dbos = new ArrayList<>(safeCountOf(cursor));

            if (Objects.nonNull(cursor)) {
                while (cursor.moveToNext()) {
                    if (emitter.isDisposed()) {
                        break;
                    }

                    int id = cursor.getInt(cursor.getColumnIndexOrThrow(BaseColumns._ID));
                    String title = cursor.getString(cursor.getColumnIndexOrThrow(CountriesColumns.NAME));
                    dbos.add(new CountryEntity().set(id, title));
                }

                cursor.close();
            }

            emitter.onSuccess(dbos);
        });
    }
}

File: app/src/main/kotlin/dev/ragnarok/fenrir/db/impl/DatabaseStorage.kt
package dev.ragnarok.fenrir.db.impl

import android.content.ContentProviderOperation
import android.content.ContentValues
import android.provider.BaseColumns
import dev.ragnarok.fenrir.db.MessengerContentProvider
import dev.ragnarok.fenrir.db.MessengerContentProvider.Companion.getCountriesContentUriFor
import dev.ragnarok.fenrir.db.column.CountriesColumns
import dev.ragnarok.fenrir.db.interfaces.IDatabaseStore
import dev.ragnarok.fenrir.db.model.entity.CountryDboEntity
import dev.ragnarok.fenrir.getInt
import dev.ragnarok.fenrir.getString
import dev.ragnarok.fenrir.util.Utils.safeCountOf
import io.reactivex.rxjava3.core.Completable
import io.reactivex.rxjava3.core.CompletableEmitter
import io.reactivex.rxjava3.core.Single
import io.reactivex.rxjava3.core.SingleEmitter

class DatabaseStorage internal constructor(base: AppStorages) : AbsStorage(base), IDatabaseStore {
    override fun storeCountries(accountId: Int, dbos: List<CountryDboEntity>): Completable {
        return Completable.create { emitter: CompletableEmitter ->
            val uri = getCountriesContentUriFor(accountId)
            val operations = ArrayList<ContentProviderOperation>(dbos.size + 1)
            operations.add(ContentProviderOperation.newDelete(uri).build())
            for (dbo in dbos) {
                val cv = ContentValues()
                cv.put(BaseColumns._ID, dbo.id)
                cv.put(CountriesColumns.NAME, dbo.title)
                operations.add(
                    ContentProviderOperation.newInsert(uri)
                        .withValues(cv)
                        .build()
                )
            }
            contentResolver.applyBatch(MessengerContentProvider.AUTHORITY, operations)
            emitter.onComplete()
        }
    }

    override fun getCountries(accountId: Int): Single<List<CountryDboEntity>> {
        return Single.create { emitter: SingleEmitter<List<CountryDboEntity>> ->
            val uri = getCountriesContentUriFor(accountId)
            val cursor = contentResolver.query(uri, null, null, null, null)
            val dbos: MutableList<CountryDboEntity> = ArrayList(safeCountOf(cursor))
            if (cursor != null) {
                while (cursor.moveToNext()) {
                    if (emitter.isDisposed) {
                        break
                    }
                    val id = cursor.getInt(BaseColumns._ID)
                    val title =
                        cursor.getString(CountriesColumns.NAME)
                    dbos.add(CountryDboEntity().set(id, title))
                }
                cursor.close()
            }
            emitter.onSuccess(dbos)
        }
    }
}